flink-runtime-web/src/main/java/org/apache/flink/runtime/webmonitor/handlers/TaskManagerLogHandler.java

```diff
@@ -0,0 +1,345 @@
+					KeepAliveWrite.flush(ctx, routed.request(), redirect);
+				} else {
+					respondAsLeader(ctx, request, routed.pathParams(), jobManager.get()._1());
+				}
+			} else {
+				KeepAliveWrite.flush(ctx, routed.request(), HandlerRedirectUtils.getUnavailableResponse());
+			}
+		} else {
+			KeepAliveWrite.flush(ctx, routed.request(), HandlerRedirectUtils.getUnavailableResponse());
+		}
+	}
+
+	/**
+	 * Response when running with leading JobManager.
+	 */
+	private void respondAsLeader(final ChannelHandlerContext ctx, final HttpRequest request, final Map<String, String> pathParams, final ActorGateway jobManager) throws Exception {
+		final String taskManagerID = pathParams.get(TaskManagersHandler.TASK_MANAGER_ID_KEY);
+
+		boolean fetch;
+		synchronized (lastRequestCompleted) {
+			if (!lastRequestCompleted.containsKey(taskManagerID)) {
+				lastRequestCompleted.put(taskManagerID, true);
+			}
+			fetch = lastRequestCompleted.get(taskManagerID);
+			lastRequestCompleted.put(taskManagerID, false);
+		}
+
+		//fetch taskmanager logs if no other process is currently doing it
+		if (fetch) {
+			//get taskmanager gateway
+			InstanceID instanceID = new InstanceID(StringUtils.hexStringToByte(taskManagerID));
+			Future<Object> future = jobManager.ask(new JobManagerMessages.RequestTaskManagerInstance(instanceID), timeout);
+			JobManagerMessages.TaskManagerInstance instance = (JobManagerMessages.TaskManagerInstance) Await.result(future, timeout);
+			Instance taskManager = instance.instance().get();
+
+			Future<Object> isBlobServiceDefined = taskManager.getActorGateway().ask(TaskManagerMessages.getIsBlobServiceDefined(), timeout);
+
+			if (!(Boolean) Await.result(isBlobServiceDefined, timeout)) {
+				display(ctx, request, "BlobService unavailable, cannot upload TaskManager logs.");
+				return;
+			}
+
+			//send log request to taskmanager
+			Future<Object> blobKeyFuture = taskManager.getActorGateway().ask(serveLogFile ? TaskManagerMessages.getRequestTaskManagerLog() : TaskManagerMessages.getRequestTaskManagerStdout(), timeout);
+			blobKeyFuture.onComplete(new OnComplete<Object>() {
+				@Override
+				public void onComplete(Throwable failure, Object success) throws Throwable {
+					try {
+						BlobKey blobKey = (BlobKey) success;
+
+						//delete previous log file, if it is different than the current one
+						if ((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).containsKey(taskManagerID)) {
+							if (!blobKey.equals((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).get(taskManagerID))) {
+								jobManager.tell(JobManagerMessages.getDeleteTaskManagerLog((serveLogFile ? lastSubmittedLog : lastSubmittedStdout).remove(taskManagerID)));
+								(serveLogFile ? lastSubmittedLog : lastSubmittedStdout).put(taskManagerID, blobKey);
+							}
+						} else {
+							(serveLogFile ? lastSubmittedLog : lastSubmittedStdout).put(taskManagerID, blobKey);
+						}
```

Comment:
I think this code block would be a bit easier to read if we refrained from using too much nesting and ternary operators.